# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: devtools/server/actors/targets/worker.js
# Commit: 6eaa4d12b58c
# Full Hash: 6eaa4d12b58cb4762729fdf53511450926383db8
# Author: Nicolas Chevobbe <nchevobbe@mozilla.com>
# Date: 2020-10-15 09:40:06
# Regressor Bug: 1633712
# File Overlap Count: 1
# Description:
#   Bug 1633712 - [devtools] Create WorkerTargets as soon as possible. r=ochameau,devtools-backward-compat-reviewers.
#   
#   This patch adds support for dedicated worker targets in the Watcher actor.
#   Shared and Service workers are not handled yet.
#   
# ==============================================================================

diff -r 10f97f47c883 -r 6eaa4d12b58c devtools/server/actors/targets/worker.js
--- a/devtools/server/actors/targets/worker.js	Thu Oct 15 05:11:58 2020 +0000
+++ b/devtools/server/actors/targets/worker.js	Thu Oct 15 05:17:52 2020 +0000
@@ -14,6 +14,8 @@
 const makeDebuggerUtil = require("devtools/server/actors/utils/make-debugger");
 const { TabSources } = require("devtools/server/actors/utils/TabSources");
 
+const Resources = require("devtools/server/actors/resources/index");
+
 exports.WorkerTargetActor = ActorClassWithSpec(workerTargetSpec, {
   targetType: Targets.TYPES.WORKER,
 
@@ -22,12 +24,18 @@
    *
    * @param {DevToolsServerConnection} connection: The connection to the client.
    * @param {WorkerGlobalScope} workerGlobal: The worker global.
+   * @param {Object} workerDebuggerData: The worker debugger information
+   * @param {String} workerDebuggerData.id: The worker debugger id
+   * @param {String} workerDebuggerData.url: The worker debugger url
+   * @param {String} workerDebuggerData.type: The worker debugger type
    */
-  initialize: function(connection, workerGlobal) {
+  initialize: function(connection, workerGlobal, workerDebuggerData) {
     Actor.prototype.initialize.call(this, connection);
 
     // workerGlobal is needed by the console actor for evaluations.
     this.workerGlobal = workerGlobal;
+
+    this._workerDebuggerData = workerDebuggerData;
     this._sources = null;
 
     this.makeDebugger = makeDebuggerUtil.bind(null, {
@@ -36,6 +44,8 @@
       },
       shouldAddNewGlobalAsDebuggee: () => true,
     });
+
+    this.notifyResourceAvailable = this.notifyResourceAvailable.bind(this);
   },
 
   form() {
@@ -43,10 +53,18 @@
       actor: this.actorID,
       threadActor: this.threadActor?.actorID,
       consoleActor: this._consoleActor?.actorID,
+      id: this._workerDebuggerData.id,
+      type: this._workerDebuggerData.type,
+      url: this._workerDebuggerData.url,
+      traits: {},
     };
   },
 
   attach() {
+    if (this.threadActor) {
+      return;
+    }
+
     // needed by the console actor
     this.threadActor = new ThreadActor(this, this.workerGlobal);
 
@@ -77,4 +95,58 @@
     // This isn't an RDP event and is only listened to from startup/worker.js.
     this.emit("worker-thread-attached");
   },
+
+  addWatcherDataEntry(type, entries) {
+    if (type == "resources") {
+      return this._watchTargetResources(entries);
+    }
+
+    return Promise.resolve();
+  },
+
+  removeWatcherDataEntry(type, entries) {
+    if (type == "resources") {
+      return this._unwatchTargetResources(entries);
+    }
+
+    return Promise.resolve();
+  },
+
+  /**
+   * These two methods will create and destroy resource watchers
+   * for each resource type. This will end up calling `notifyResourceAvailable`
+   * whenever new resources are observed.
+   */
+  _watchTargetResources(resourceTypes) {
+    return Resources.watchResources(this, resourceTypes);
+  },
+
+  _unwatchTargetResources(resourceTypes) {
+    return Resources.unwatchResources(this, resourceTypes);
+  },
+
+  /**
+   * Called by Watchers, when new resources are available.
+   *
+   * @param Array<json> resources
+   *        List of all available resources. A resource is a JSON object piped over to the client.
+   *        It may contain actor IDs, actor forms, to be manually marshalled by the client.
+   */
+  notifyResourceAvailable(resources) {
+    this.emit("resource-available-form", resources);
+  },
+
+  destroy() {
+    Actor.prototype.destroy.call(this);
+
+    if (this._sources) {
+      this._sources.destroy();
+      this._sources = null;
+    }
+
+    this.workerGlobal = null;
+    this._dbg = null;
+    this.threadActor = null;
+    this._consoleActor = null;
+  },
 });